File: client/src/sim/Sim.ts
/**
 * SimState + the fixed-order tick function. This is the whole deterministic
 * core: same seed + same per-tick inputs ⇒ byte-identical state hashes on
 * every client.
 *
 * NOTE (deviation from the original plan): there is no `shells[]` array —
 * trap state is folded into PlayerState (see Shell.ts for the rationale).
 *
 * Tick system order (determinism contract — do not reorder):
 *  1. players: resolve buffered input & move (array order);
 *  2. bomb placement from action edges (array order);
 *  3. fuse decrement; detonations + chains → explosion cells, soft-brick
 *     destruction, item drops (PRNG draws — see Explosion.ts);
 *  4. explosion-cell hits: trap players (array order);
 *  5. item pickups (item-array order; first player in array order wins);
 *  6. shells: rescue, then timers → elimination;
 *  7. age & cull explosion cells;
 *  8. win check → maybe phase = OVER (always PvP: OVER once <=1 distinct team
 *     has an alive player);
 *  9. recompute stateHash, tick + 1.
 *
 * PRNG draw order across a whole tick: step 3 (drop rolls) only.
 * `createInitialState` draws: map generation ONLY (no per-entity spawn draws).
 */
import { MATCH_MAX_TICKS } from '../../../shared/constants';
import { GamePhase } from '../../../shared/types';
import {
  type FeelParams,
  cornerAssistMt,
  inputBufferTicks,
  moveSpeedMt,
} from '../config/FeelParams';
import { type BombState, bombPressedEdge, tryPlaceBomb } from './Bomb';
import {
  type ExplosionState,
  explosionAt,
  processDetonations,
} from './Explosion';
import { hashSimState } from './Hash';
import { type InputFrame, NO_INPUT } from './InputBuffer';
import { type ItemState, applyItem } from './Item';
import { type MapKind, SPAWN_CORNERS, type TileGrid, generateMap } from './Map';
import {
  type PlayerState,
  type SimParams,
  clonePlayer,
  createPlayer,
  stepPlayerMovement,
  tileOf,
} from './Player';
import { stepShells, trapPlayer } from './Shell';

export type { SimParams } from './Player';

export interface SimState {
  readonly tick: number;
  readonly phase: GamePhase;
  /** Mulberry32 state (uint32), threaded through every random draw. */
  readonly prng: number;
  /** Integer params derived once from FeelParams at match start. */
  readonly params: SimParams;
  readonly map: TileGrid;
  /**
   * Which map layout this match uses (classic / pirate …). A whole-match
   * constant fixed at createInitialState and carried forward unchanged every
   * tick — deliberately NOT hashed (exactly like `params.pvp` / `team`): the
   * grid TILES it selects are hashed via map generation, the selector itself is
   * a non-hashed match constant. Bots read it to pick a per-map decision profile.
   */
  readonly mapKind: MapKind;
  readonly players: readonly PlayerState[];
  readonly bombs: readonly BombState[];
  readonly explosions: readonly ExplosionState[];
  readonly items: readonly ItemState[];
  /** Canonical FNV-1a hash of everything above (see Hash.ts). */
  readonly stateHash: number;
}

/**
 * Create the initial match state.
 *
 * PRNG order: `generateMap` (see Map.ts) — and nothing else. There are no
 * per-entity spawn draws, so the whole initial PRNG state is fixed by the map.
 *
 * `opts` is optional. The `team`/`pvp`/`map` fields are deliberately NOT
 * hashed: they are whole-match constants fixed here once and (in net mode)
 * carried identically to every client by MatchStart. The map kind selects the
 * grid layout (whose tiles ARE hashed via map generation); the selector itself
 * is a match constant like teams. Team default = slot (teams[i] ?? i).
 */
export function createInitialState(
  seed: number,
  feelParams: FeelParams,
  numPlayers: number,
  opts?: {
    pvp?: boolean; // 預設 false（vestigial：行為一律 PvP，見 SimParams.pvp）
    teams?: readonly number[]; // 預設 teams[i] = i（隊伍即 slot）
    map?: MapKind; // 預設 'classic'（地圖布局種類，整場固定，非隨機）
  },
): SimState {
  // pvp is vestigial — kept on SimParams but never branches anything; the win
  // condition is always last-team-standing. Stored for compatibility only.
  const pvp = opts?.pvp ?? false;
  const teams = opts?.teams;

  const mapKind: MapKind = opts?.map ?? 'classic';
  const [map, prng] = generateMap(seed >>> 0, mapKind);

  const params: SimParams = Object.freeze({
    moveSpeedMt: moveSpeedMt(feelParams),
    cornerAssistMt: cornerAssistMt(feelParams),
    inputBufferTicks: inputBufferTicks(feelParams),
    // pvp is a whole-match constant; NOT hashed (see SimParams in Player.ts).
    pvp,
  });

  const n = Math.max(1, Math.min(numPlayers, SPAWN_CORNERS.length));
  const players: PlayerState[] = [];
  for (let i = 0; i < n; i++) {
    const corner = SPAWN_CORNERS[i];
    // team is a whole-match constant; NOT hashed (see PlayerState in Player.ts).
    // Default team = slot (each player on its own team unless opts.teams given).
    if (corner !== undefined) {
      players.push(createPlayer(i, corner[0], corner[1], teams?.[i] ?? i));
    }
  }

  const state: SimState = {
    tick: 0,
    phase: GamePhase.PLAYING,
    prng,
    params,
    map,
    mapKind,
    players,
    bombs: [],
    explosions: [],
    items: [],
    stateHash: 0,
  };
  return { ...state, stateHash: hashSimState(state) };
}

/**
 * Advance the simulation one tick. Pure: never mutates `state`; returns a
 * new SimState. `inputs[i]` is the InputFrame for players[i] (missing ⇒
 * NO_INPUT). System order documented at the top of this file.
 */
export function tick(state: SimState, inputs: readonly InputFrame[]): SimState {
  if (state.phase !== GamePhase.PLAYING) {
    // Frozen end state: only the tick counter advances.
    const frozen: SimState = { ...state, tick: state.tick + 1 };
    return { ...frozen, stateHash: hashSimState(frozen) };
  }

  let prng = state.prng;
  const grid = new Uint8Array(state.map);
  const players = state.players.map(clonePlayer);
  let bombs = state.bombs.map((b) => ({ ...b }));
  let explosions = state.explosions.map((c) => ({ ...c }));
  let items = state.items.map((it) => ({ ...it }));

  // (1) players: input resolution + movement.
  for (let i = 0; i < players.length; i++) {
    const pl = players[i];
    if (pl === undefined) continue;
    stepPlayerMovement(grid, bombs, pl, inputs[i] ?? NO_INPUT, state.params);
  }

  // (2) bomb placement on action rising edge.
  for (let i = 0; i < players.length; i++) {
    const pl = players[i];
    if (pl === undefined) continue;
    const input = inputs[i] ?? NO_INPUT;
    if (bombPressedEdge(pl.prevAction, input.action)) {
      const bomb = tryPlaceBomb(bombs, pl, tileOf(pl.posX), tileOf(pl.posY));
      if (bomb !== null) bombs.push(bomb);
    }
    pl.prevAction = input.action;
  }

  // (3) fuses + detonations (+ chains, brick destruction, item drops).
  for (const b of bombs) b.fuseTicks -= 1;
  const det = processDetonations(grid, bombs, prng);
  bombs = det.bombs;
  prng = det.prng;
  explosions.push(...det.cells);
  // A second (new) detonation destroys items already lying on the floor; the
  // items that just dropped THIS tick are not touched by this tick's own
  // explosion (the first bomb reveals the item, a later bomb burns it). Filter
  // existing items against this tick's new cells BEFORE pushing det.items.
  if (det.cells.length > 0 && items.length > 0) {
    items = items.filter((it) => !explosionAt(det.cells, it.tileX, it.tileY));
  }
  items.push(...det.items);
  for (const slot of det.detonatedOwners) {
    const owner = players.find((p) => p.slot === slot);
    if (owner !== undefined && owner.activeBombs > 0) owner.activeBombs -= 1;
  }

  // (4) explosion-cell hits: trap players.
  if (explosions.length > 0) {
    for (const pl of players) {
      if (
        pl.alive &&
        !pl.trapped &&
        explosionAt(explosions, tileOf(pl.posX), tileOf(pl.posY))
      ) {
        trapPlayer(pl);
      }
    }
  }

  // (5) item pickups.
  if (items.length > 0) {
    const remaining: ItemState[] = [];
    for (const it of items) {
      let taken = false;
      for (const pl of players) {
        if (
          pl.alive &&
          !pl.trapped &&
          tileOf(pl.posX) === it.tileX &&
          tileOf(pl.posY) === it.tileY
        ) {
          applyItem(pl, it.kind);
          taken = true;
          break;
        }
      }
      if (!taken) remaining.push(it);
    }
    items = remaining;
  }

  // (6) shells: rescue, then timers → elimination.
  stepShells(players);

  // (7) age & cull explosion cells.
  for (const c of explosions) c.ttlTicks -= 1;
  explosions = explosions.filter((c) => c.ttlTicks > 0);

  // (8) win check (always PvP / last-team-standing). The resulting phase IS
  // hashed. OVER once at most one distinct team still has an alive player
  // (a single survivor wins; everyone gone is also OVER), OR once the hard
  // match time cap is hit — the surviving teams are then resolved outside the
  // sim by most-survivors → item tiebreak → draw (see Outcome.ts).
  const nextTick = state.tick + 1;
  const aliveTeams = new Set<number>();
  for (const p of players) if (p.alive) aliveTeams.add(p.team);
  const phase =
    aliveTeams.size <= 1 || nextTick >= MATCH_MAX_TICKS
      ? GamePhase.OVER
      : GamePhase.PLAYING;

  // (9) assemble + hash.
  const next: SimState = {
    tick: nextTick,
    phase,
    prng,
    params: state.params,
    map: grid,
    // mapKind is a non-hashed whole-match constant — carry it forward verbatim.
    mapKind: state.mapKind,
    players,
    bombs,
    explosions,
    items,
    stateHash: 0,
  };
  return { ...next, stateHash: hashSimState(next) };
}

File: rl/env/sim.py
# rl/env/sim.py
"""
SimState + fixed-order tick(). Ported from client/src/sim/Sim.ts.
Tick system order (do NOT reorder — determinism contract):
  1. players: resolve input & move
  2. bomb placement from action edges
  3. fuse decrement; detonations → explosion cells, brick destruction, item drops
  4. explosion-cell hits: trap players
  5. item pickups
  6. shells: rescue then timers → elimination
  7. age & cull explosion cells
  8. sudden death
  9. win check → phase OVER
 10. tick + 1
"""
import numpy as np
from rl.env.constants import (
    MILLITILE, FUSE_TICKS, SPARK_TICKS, MATCH_MAX_TICKS,
    MAP_COLS, MAP_ROWS, PUSH_CHARGE_TICKS,
    DEFAULT_MOVE_SPEED, DEFAULT_CORNER_ASSIST, DEFAULT_INPUT_BUFFER_MS, TICK_HZ,
)
from rl.env.types import (
    PHASE_PLAYING, PHASE_OVER,
    DIR_NONE, DIR_UP, DIR_DOWN, DIR_LEFT, DIR_RIGHT,
    ACTION_BOMB,
    InputFrame, NO_INPUT,
    TILE_EMPTY, TILE_HARD, TILE_PUSH,
)
from rl.env.map_gen import generate_map, map_spawns
from rl.env.player import (
    create_player, step_entity, tile_of,
    player_speed_mt_per_tick,
)
from rl.env.bomb import bomb_at, bomb_pressed_edge, try_place_bomb
from rl.env.explosion import process_detonations, explosion_covers
from rl.env.item import apply_item
from rl.env.shell import step_shells, trap_player
from rl.env.sudden_death import step_sudden_death

_MOVE_SPEED_MT = round(DEFAULT_MOVE_SPEED * MILLITILE)       # 5000
_CORNER_ASSIST_MT = round(DEFAULT_CORNER_ASSIST * MILLITILE)  # 250
_INPUT_BUFFER_TICKS = round(DEFAULT_INPUT_BUFFER_MS / (1000 / TICK_HZ))  # 7


def create_initial_state(seed: int, map_kind: str = 'classic', num_players: int = 4) -> dict:
    grid, prng = generate_map(seed, map_kind)
    spawns = map_spawns(map_kind)
    players = []
    for i in range(num_players):
        sx, sy = spawns[i % len(spawns)]
        players.append(create_player(slot=i, tile_x=sx, tile_y=sy, team=i))
    return {
        'tick': 0,
        'phase': PHASE_PLAYING,
        'prng': prng,
        'map_kind': map_kind,
        'grid': grid,
        'players': players,
        'bombs': [],
        'explosions': [],
        'items': [],
    }


def _is_open(grid: np.ndarray, bombs: list[dict], x: int, y: int) -> bool:
    if x < 0 or x >= MAP_COLS or y < 0 or y >= MAP_ROWS:
        return False
    if grid[y * MAP_COLS + x] != TILE_EMPTY:
        return False
    return bomb_at(bombs, x, y) is None


def _dir_dx(d: int) -> int:
    return -1 if d == DIR_LEFT else (1 if d == DIR_RIGHT else 0)


def _dir_dy(d: int) -> int:
    return -1 if d == DIR_UP else (1 if d == DIR_DOWN else 0)


def can_push(
    grid: np.ndarray, bombs: list[dict],
    pos_x: int, pos_y: int, direction: int,
) -> bool:
    """True if a PUSH crate is directly ahead and can be shoved.
    Requires player to be exactly tile-centred on both axes.
    Mirrors canPush in client/src/sim/Player.ts."""
    if pos_x % MILLITILE != 0 or pos_y % MILLITILE != 0:
        return False
    dx = _dir_dx(direction)
    dy = _dir_dy(direction)
    if dx == 0 and dy == 0:
        return False
    cx = tile_of(pos_x)
    cy = tile_of(pos_y)
    ax, ay = cx + dx, cy + dy   # tile directly ahead (the crate)
    bx, by = ax + dx, ay + dy   # tile beyond (where crate slides to)
    if not (0 <= ax < MAP_COLS and 0 <= ay < MAP_ROWS):
        return False
    if grid[ay * MAP_COLS + ax] != TILE_PUSH:
        return False
    return _is_open(grid, bombs, bx, by)


def apply_push(grid: np.ndarray, pos_x: int, pos_y: int, direction: int) -> None:
    """Slide the PUSH crate one tile ahead. MUTATES grid.
    Caller must have verified can_push this tick.
    Mirrors applyPush in client/src/sim/Player.ts."""
    dx = _dir_dx(direction)
    dy = _dir_dy(direction)
    ax = tile_of(pos_x) + dx
    ay = tile_of(pos_y) + dy
    grid[ay * MAP_COLS + ax] = TILE_EMPTY
    grid[(ay + dy) * MAP_COLS + (ax + dx)] = TILE_PUSH


def _copy_state(state: dict) -> dict:
    """Shallow-copy state, manually copying each mutable field.

    Avoids copy.deepcopy overhead (~3× faster) while staying correct:
    - grid: numpy array — use ndarray.copy()
    - players/bombs/explosions/items: list of flat dicts — copy each dict
    - held_stack inside player: a list that is never mutated by the sim,
      but copy it to avoid any shared-reference surprises
    - prng / tick / phase / map_kind: scalars / str — no copy needed
    """
    s: dict = {
        'tick': state['tick'],
        'phase': state['phase'],
        'prng': state['prng'],
        'map_kind': state['map_kind'],
        'grid': state['grid'].copy(),
        'players': [
            {**p, 'held_stack': list(p['held_stack'])}
            for p in state['players']
        ],
        'bombs': [dict(b) for b in state['bombs']],
        'explosions': [dict(e) for e in state['explosions']],
        'items': [dict(i) for i in state['items']],
    }
    return s


def tick(state: dict, inputs: list[InputFrame]) -> dict:
    """Advance state by one tick. Returns a NEW state (does not mutate input)."""
    s = _copy_state(state)
    grid = s['grid']
    players = s['players']
    bombs = s['bombs']
    explosions = s['explosions']
    items = s['items']
    prng = s['prng']

    # ── Step 1: resolve input & move ─────────────────────────────────────────
    for i, p in enumerate(players):
        if not p['alive'] or p['trapped']:
            p['push_charge_dir'] = 0
            p['push_charge_ticks'] = 0
            continue
        inp = inputs[i] if i < len(inputs) else NO_INPUT
        speed = player_speed_mt_per_tick(_MOVE_SPEED_MT, p['speed_bonus_tenths'])

        def open_fn(ax, bx, _g=grid, _b=bombs):
            return _is_open(_g, _b, ax, bx)

        d = inp.dir
        charging = False
        if d != DIR_NONE:
            nx, ny, moved = step_entity(open_fn, p['pos_x'], p['pos_y'], d, speed, _CORNER_ASSIST_MT)
            if moved:
                p['pos_x'] = nx
                p['pos_y'] = ny
                p['facing'] = d
            elif can_push(grid, bombs, p['pos_x'], p['pos_y'], d):
                p['facing'] = d
                charging = True
                p['push_charge_ticks'] = (
                    p['push_charge_ticks'] + 1 if p['push_charge_dir'] == d else 1
                )
                p['push_charge_dir'] = d
                if p['push_charge_ticks'] >= PUSH_CHARGE_TICKS:
                    apply_push(grid, p['pos_x'], p['pos_y'], d)
                    p['push_charge_ticks'] = 0
                    p['push_charge_dir'] = 0
        if not charging:
            p['push_charge_dir'] = 0
            p['push_charge_ticks'] = 0

    # Step 1½: a pushed crate may now sit on a floor item — remove that item.
    # (grid was mutated above; item-tile==PUSH ⟺ a crate just slid onto it.)
    if items:
        items = [it for it in items if grid[it['tile_y'] * MAP_COLS + it['tile_x']] != TILE_PUSH]

    # ── Step 2: bomb placement ────────────────────────────────────────────────
    new_bombs = list(bombs)
    for i, p in enumerate(players):
        inp = inputs[i] if i < len(inputs) else NO_INPUT
        if bomb_pressed_edge(p['prev_action'], inp.action):
            tx = tile_of(p['pos_x'])
            ty = tile_of(p['pos_y'])
            b = try_place_bomb(new_bombs, p, tx, ty)
            if b is not None:
                new_bombs.append(b)
        p['prev_action'] = inp.action

    # ── Step 3: fuse decrement + detonations ─────────────────────────────────
    for b in new_bombs:
        b['fuse_ticks'] -= 1

    result = process_detonations(grid, new_bombs, prng)
    bombs_after = result['bombs']
    new_cells = result['cells']
    new_items = result['items']
    prng = result['prng']
    # Credit detonated owners
    detonated_owners = result['detonated_owners']
    owner_counts: dict[int, int] = {}
    for owner in detonated_owners:
        owner_counts[owner] = owner_counts.get(owner, 0) + 1
    for p in players:
        if p['slot'] in owner_counts:
            p['active_bombs'] = max(0, p['active_bombs'] - owner_counts[p['slot']])

    all_cells = [c for c in explosions if c['ttl_ticks'] > 0] + new_cells

    # ── Step 4: explosion hits → trap players ────────────────────────────────
    for p in players:
        if not p['alive'] or p['trapped']:
            continue
        if explosion_covers(all_cells, p['pos_x'], p['pos_y']):
            trap_player(p)

    # ── Step 5: item pickups ──────────────────────────────────────────────────
    items_remaining = list(items) + new_items
    items_after = []
    for it in items_remaining:
        claimed = False
        for p in players:
            if p['alive'] and not p['trapped']:
                if tile_of(p['pos_x']) == it['tile_x'] and tile_of(p['pos_y']) == it['tile_y']:
                    apply_item(p, it['kind'])
                    claimed = True
                    break
        if not claimed:
            items_after.append(it)

    # ── Step 6: shells ────────────────────────────────────────────────────────
    step_shells(players)

    # ── Step 7: age & cull explosion cells ───────────────────────────────────
    explosions_after = []
    for c in all_cells:
        c['ttl_ticks'] -= 1
        if c['ttl_ticks'] > 0:
            explosions_after.append(c)

    # ── Step 8: sudden death ──────────────────────────────────────────────────
    step_sudden_death(grid, players, s['tick'] + 1)  # applies at new tick

    # ── Step 9: win check ─────────────────────────────────────────────────────
    alive_teams = {p['team'] for p in players if p['alive']}
    if len(alive_teams) <= 1 or s['tick'] + 1 >= MATCH_MAX_TICKS:
        phase = PHASE_OVER
    else:
        phase = PHASE_PLAYING

    return {
        'tick': s['tick'] + 1,
        'phase': phase,
        'prng': prng,
        'map_kind': s['map_kind'],
        'grid': grid,
        'players': players,
        'bombs': bombs_after,
        'explosions': explosions_after,
        'items': items_after,
    }
